### ISSUE TITLE
Ease generator understanding

### ISSUE THREAD

**Issue by Tcharl** [jhipster/generator-jhipster#14235]
##### **Overview of the feature request**

Not really a request but more a plan for what I'm trying to currently achieve.

The plan is divided into multiple steps and I think we gonna win some things at each of them.

1. Use constants in all js files.
As you may have noticed, I'm currently introducing options constants anywhere I can in the generators js files, replacing `buildTool == 'maven'` by `buildTool == MAVEN` and this for all constant.
As a developer with an IDE, I love the folding capability to see where a particular constant is used.
Using constants instead of plain string will help to identify all the generators, needles and helpers that uses a jhipster option.
Also, I think it will help the test maintenance and the removal of some jdl historic syntax mismatch (for example the `false`vs the `no`).

2. Align the generator's context feeding

Context is populated by options differently on each generator: 
for example in the [_loading()](https://github.com/jhipster/generator-jhipster/blob/e22ca7cfd00f2a905ae7996d7adfcf82096bb4e5/generators/client/index.js#L163) method in the client generator whereas its in the [getConfig()](https://github.com/jhipster/generator-jhipster/blob/e22ca7cfd00f2a905ae7996d7adfcf82096bb4e5/generators/ci-cd/index.js#L90) one in the ci-cd generator.

**Question**: Can someone tell me which is the generator that uses the best practices so that I'll replicate it in all of them? I'm not a Yeoman black belt, and I'm afraid to do it wrong.

So that we'll have something that is replicable and begin to mutualize some methods names and calls. Also it will be easier for everyone to have a standard canvas/template for method override and extension (I'm thinking of blueprints).

3. Use 'derived properties' in templates instead of using logic (equality tests, string tests)

The final step would be to use only boolean logic in templates, thus injecting derived attributes in generators context:
for example, injecting a `ìsBuildToolMaven` instead of using `buildTool == 'maven'`in all the ejs.
This last step will bring additional value on top of the two previous one (and can only be achieved well with the two previous one implemented):

- Easy folding: a search in the IDE will lead to all check achieved on an option
- Templates will be a little bit cleaner and just be dumb as possible (only containing OR and AND ejs checks)
- Also, any newcomer will be able to see where using the 'maven' 'buildTool' option is impacting the generation, so it will be easier for him to either add a similar option (sbt, bnd, ant :-p) or to override some option's behaviour.
- Finally, it should be way easier for the core committer to maintain an option, and may be to extract them in their dedicated templates (I would love to have template composition depending on options one day).

##### **Motivation for or Use Case**

Well known and adopted frameworks are very clean at their core (Spring, Hibernate) so that they can scale and adapt more easily.

##### **Related issues or PR**

[14206](https://github.com/jhipster/generator-jhipster/pull/14206)
[14203](https://github.com/jhipster/generator-jhipster/pull/14203)

Last request: can you please be reactive on MR contributing to this request please?
This is a huge work and impacting all the repo's file: there's a high risk each time to face many merge conflict and also no one would want to review all the files of the generator at once: it would be better to accept it by small increment...

If someone else want to contribute to the initiative, please come! There's easily a room for many of us (also, if new PR could anticipate, \o/).

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by pascalgrimaud** [comment]
My idea for v8 would be to not have these conditions, at all.
Using an Hexagonal Architecture will help.

Each files will be designed for each technologies, very few/no common templates.

For example, look at https://github.com/jhipster/generator-jhipster/blob/main/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs

Do you think it's easy to maintain https://github.com/jhipster/generator-jhipster/blob/main/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs#L1540-L1557 ?

I don't think so :)

So my opinion will be to not put so much effort on this.
I'll create a new ticket for v8, to discuss about the future of JHipster

**Comment by mshima** [comment]
> Context is populated by options differently on each generator:
> for example in the [_loading()](https://github.com/jhipster/generator-jhipster/blob/e22ca7cfd00f2a905ae7996d7adfcf82096bb4e5/generators/client/index.js#L163) method in the client generator whereas its in the [getConfig()](https://github.com/jhipster/generator-jhipster/blob/e22ca7cfd00f2a905ae7996d7adfcf82096bb4e5/generators/ci-cd/index.js#L90) one in the ci-cd generator.

You should try to follow the main generators (applications generators) practices, but ci-cd and other generators use cases are quite different.
They are supposed to be executed on an already configured application, so app configurations should be ready at initializing because they will be required for prompts.

> **Question**: Can someone tell me which is the generator that uses the best practices so that I'll replicate it in all of them? I'm not a Yeoman black belt, and I'm afraid to do it wrong.

- I would recommend to stop loading/saving configurations into the generator and use directly at `this.jhipsterConfig.*`.
It is realtime read/write and removes a lot of unnecessary roundtrips (`config -> generator -> config`).
- Prefer tasks implementation in this order:
  - sync
  - async/await
  - return Promise
  - `const done = this.async(); done();`
- Use `this.spawnCommand/this.spawnCommandAsync` instead of shelljs/other helpers. The built-in methods sets destinationPath as cwd.

> So my opinion will be to not put so much effort on this.

Unless JHipster will move away from yeoman, this is a good cleanup.

**Comment by Tcharl** [comment]
> Each files will be designed for each technologies, very few/no common templates.
That was I thought when saying
> I would love to have template composition depending on options one day

**Comment by pascalgrimaud** [comment]
Yes, so we agree :-)
Anyway, if I'm not wrong, there are no breaking change (for release v7), and it can be done steps by steps, right ?
In this case, plz, go ahead, but with smallest PR as possible, so it will be easy to review

**Comment by Tcharl** [comment]
@pascalgrimaud you're right, it should not introduce any breaking change. The only ones that could come would be if we modify some of the generators context feeding logic or sequence (or cleanup some hacky things), but we'll see if it happens and discuss about it when the according PRs will spawn

**Comment by Tcharl** [comment]
Quick question on this story: should we bind constants between jdl and jsons/yo-rc when we can or keep them splitted in all cases?

For example, we can bind:  `baseName` in jdl as well as in yo-rc. But we cannot (or at least not for now `ManyToMany` in jdl and `many-to-many` in `entity.json`

This has pro and cons:
- For sure, binding can simplify the overall generator + jdl code: we'll be able to merge some method, and the overall understanding for contributors will be easier. Also for end users (they'll see the same thing in their JDL, their configs, and maybe their code if we choose the right keyword.

- On the other side, we'll have less flexibility in the terms we'll choose and the change we'll be able to do (the grammar and configs will be more fixed).

**Comment by MathieuAA** [comment]
Binding is coupling, and coupling decreases the opportunities to have a choice, so... If the name are the same, great, but this shouldn't be a rule.

**Comment by DanielFran** [comment]
Keep open

**Comment by pascalgrimaud** [comment]
Adding a bounty as it is a lot of work

**Comment by Tcharl** [comment]
Thank you Pascal!

**Comment by renanfranca** [comment]
@Tcharl congratulations for the idea! This will help me a lot to understand the code base.

How can i help here? Where do you need help?

It will be my first time as Jhipster contribute 😀. I want to help and learn more about the code base too.

My experience with Jhipster: I built an personal blue print to slightly change how the spring boot Backend code is generated and add some custom annotations.

**Comment by Tcharl** [comment]
Hi @renanfranca , glad it has helped at least someone :-).

Any help is welcome for sure: I'm amending all the files from top to bottom (I'm at generator/languages). If you want to contribute, you can amend file from the bottom to the top (starting from utils/relationship.js), we'll join one day or another :-).
 
You can take a look at the MR that have already been done if you want some guidance.
Basically, the task is made of:

- Looking if the variable exists in jdl/jhipster/x-types.js
- Extending BlueprintBaseGenerator for all legacy generators and modify methods accordingly
- Removing comparison in ejs in favor of derived constants
- Removing strings in generator js files in favor of constants.

**Comment by renanfranca** [comment]
> Hi @renanfranca , glad it has helped at least someone :-).
> 
> Any help is welcome for sure: I'm amending all the files from top to bottom (I'm at generator/languages). If you want to contribute, you can amend file from the bottom to the top (starting from utils/relationship.js), we'll join one day or another :-).
> 
> You can take a look at the MR that have already been done if you want some guidance.
> Basically, the task is made of:
> 
> * **Looking if the variable exists in jdl/jhipster/x-types.js**
> * Extending BlueprintBaseGenerator for all legacy generators and modify methods accordingly
> * Removing comparison in ejs in favor of derived constants
> * Removing strings in generator js files in favor of constants.

@Tcharl, thank you for the fast and detailed response! I've already made the fork and I'm trying to understand what I have to change at utils/relationship.js. For example:


```
(...)
function _derivedProperties(relationship) {
  _.defaults(relationship, {
    relationshipOneToOne: relationship.relationshipType === 'one-to-one',
    relationshipOneToMany: relationship.relationshipType === 'one-to-many',
    relationshipManyToOne: relationship.relationshipType === 'many-to-one',
    relationshipManyToMany: relationship.relationshipType === 'many-to-many',
    otherEntityUser: relationship.otherEntityName === 'user',
  });
}
(...)
  _.defaults(relationship, {
    // let ownerSide true when type is 'many-to-one' for convenience.
    // means that this side should control the reference.
    ownerSide: relationship.ownerSide || relationship.relationshipType === 'many-to-one',
    collection: relationship.relationshipType === 'one-to-many' || relationship.relationshipType === 'many-to-many',
  });
(...)
   if (otherRelationship) {
      relationship.otherSideReferenceExists = true;
      if (
        !(relationship.relationshipType === 'one-to-one' && otherRelationship.relationshipType === 'one-to-one') &&
        !(relationship.relationshipType === 'many-to-one' && otherRelationship.relationshipType === 'one-to-many') &&
        !(relationship.relationshipType === 'one-to-many' && otherRelationship.relationshipType === 'many-to-one') &&
        !(relationship.relationshipType === 'many-to-many' && otherRelationship.relationshipType === 'many-to-many')
      ) {
(...)
```
There are those relationship-type as string at the code. I search for this types at jdl/jhipster/ folder and they aren't defined. **Question:** _Should I leave that way (no work to be done here)?_ 

Looking at next top file utils\liquibase.js:
```
(...)
function parseLiquibaseColumnType(entity, field) {
  const fieldType = field.fieldType;
  if (fieldType === 'String' || field.fieldIsEnum) {
    return `varchar(${field.fieldValidateRulesMaxlength || 255})`;
  }

  if (fieldType === 'Integer') {
    return 'integer';
  }

  if (fieldType === 'Long') {
    return 'bigint';
  }
(...)
```
Which has constants defined jdl\jhipster\application-options.js. **Question:** _Should I start to replace them, right?_

**Comment by Tcharl** [comment]
Everything that can make our lovely developper folding to be able to add a new type/technology/relationship is welcome :-)

**Comment by Tcharl** [comment]
Bounty claimed: https://opencollective.com/generator-jhipster/expenses/47720

**Comment by Tcharl** [comment]
Thank you community, it has paid my vacation close to the ocean! ❤
